bot: allow more fine-grained reply control

euank opened the pull request:
Only light manual testing done. I need to add more info in the last commit message too.

LinuxMercedes commented on bot.js:
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
+  //   pmExtra: false, // whether to PM the whole message if it overflows
+  // }
+  var customReply = function(opts, args) {
+    var maxLineChars = 460; // convenient lie; the true number is complicated due to having to account for nick-length etc
```
You might be able to get this from `node-irc`. See https://github.com/martynsmith/node-irc/blob/master/lib/irc.js#L1069

euank replied:
That's a convenient lie too I think, but I'll grab that value and call it good enough

LinuxMercedes commented on bot.js:
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
+      spamReply.apply(this, args);
+    }
+
+    var repStr = bot.stringifyArgs.apply(this, args);
```
Nit: I'd add a comment here saying something to the effect of "calculate number of messages to send these lines"?

LinuxMercedes commented on bot.js:
```diff
@@ -335,36 +343,82 @@ bot.isChannel = function(name) {
-    bot.client.action(to, repStr);
+    // Default to not spam if it's not a pm
+    args = Array.prototype.slice.call(arguments);
+    args.unshift({lines: 2, replaceNewLines: false, pmExtra: false});
```
Is 2 enough lines here for stuff like wolfram alpha?

euank replied:
I'll update wolfram and twitter to use `custom`

euank commented:
Added some tests and reworked some of the logic based on writing the tests. Still not totally sure if this is all what I want... I'll write some more tests and then convert the modules that should use this over

drusepth commented:
@euank What is the intended functionality here? Is it just some kind of throttling for how often it spits responses out to a channel?

euank commented:
@drusepth no throttling, that's a configuration the irc-client lib can already do separately.

This is integrating the t_t anti-spam mechanism (akin to [this](https://github.com/euank/reckless-irc-exec/blob/43f60e75cff11f889c8fca83421de507e8cb5bec/bot.js#L15-L42)) in preparation for merging in similarly spammy code.

The intent is that modules, by default, when they call `reply(someString)` will have that string truncated to a reasonable length unless the module explicitly intends to spew more, either via multiple `reply` calls, or via `reply.spam` or `reply.custom`.

A good example of a module this helps with now is the `!ud` module. Because that's replying with the result of network traffic based on the user query, right now it's easy to spam a giant wall of text.

This will make that impossible unless the module explicitly opts into being able to spam so much.

LinuxMercedes commented on bot.js:
```diff
@@ -335,36 +343,111 @@ bot.isChannel = function(name) {
+      // we know it doesn't fit like this already, let's say a trimmed down version.
+      let linesToSay = [];
+      let maxLength = maxLineChars * opts.lines - 4; // -4 to leave room for '...'
+      for(let i=0; i < lines.length; i++) {
```
I feel like this loop could be more clearly done as a `reduce`, but maybe that's just the haskell in my brain talking.

euank replied:
Because of the `i == (opts.lines - 1)` bit it seemed a little less clearcut.

LinuxMercedes replied:
Ahh, I see. Maybe a `map()` +`take(opts.lines)`?

(Really, though, this is just a nice bikeshed. If you're happy with it, I'm happy with it =D)
